### PR TITLE
Update the custom target example

### PR DESCRIPTION
**Pull request by JohnK1987**
This PR will update the custom target example in few ways for be more easy to understand for novices, I hope.
* description in top lvl CMakeList about basic possibilities.
* add missing files into project to be complete and working directly after clone
* update to latest MbedOS (minor)
* update readme

The main reason is in context of https://github.com/mbed-ce/mbed-os/pull/250

### CODE REVIEW

**Comment by JohnK1987** [comment]
Build tests failed because of upload methods are not on the machine. I do not know what is supposed to do with this...

**Comment by multiplemonomials** [comment]
Oh, just pass -DUPLOAD_METHOD=NONE to cmake in the CI script

**Comment by JohnK1987** [comment]
Ok, I did something but not sure I did it right :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,21 +1,77 @@
...
+#[[ Here you can set path for custom upload config .cmake (optional example) ]]
+# set(CUSTOM_UPLOAD_CFG_PATH ${CMAKE_SOURCE_DIR}/${MBED_TARGET}/${MBED_TARGET}.cmake)
+
+#[[ Note: For custom target you need also an upload method and we have few options how you can do that
+    - use the variable CUSTOM_UPLOAD_CFG_PATH below
+    - use default expected path for custom targets upload methods where you create your own config
+      MY_PROJECT/custom_targets/upload_method_cfg
+    - of course you can do it by your self dierectly via cmake in this file
+
+    For regular targets we use mbed-os/tree/master/targets/upload_method_cfg ]]
```
Comment (multiplemonomials) [inline CMakeLists.txt]: Can we link to the upload method wiki page and also to an example upload method config from Mbed CE?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,21 +1,77 @@
...
-set(CUSTOM_TARGETS_JSON_PATH custom_targets.json) # This activates the custom targets in custom_targets.json
+#### Initialize Mbed OS build system. ####
+######################################################################################################
+### Block of including .json5 files. Files of this block must be included before the app.cmake
 
+#[[ Set path of mbed_app.json (necessary everytime if mbed_app.json5 is in project) ]]
+set(MBED_APP_JSON_PATH mbed_app.json5)
+
+###--------------------------------------------------------------------------------------------------- 
+#[[ This part is intended to custom targets5 and activates the custom_targets.json5
```
Comment (multiplemonomials) [inline CMakeLists.txt]: I'm afraid I don't really understand what this comment is saying, could you rephrase?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,21 +1,77 @@
...
+
+### include app.cmake (necessary everytime) ###
 include(mbed-os/tools/cmake/app.cmake)
 
-# Include folder with the custom targets.
-# This needs to be done before add_subdirectory(mbed-os).
+######################################################################################################
+### Block of including project folders
+
+#[[ In case of custom targets must be the folder included before mbed-os subdirectory
```
Comment (multiplemonomials) [inline CMakeLists.txt]: Could we rephrase as
```cmake
# If using a custom target, the subdirectory containing the custom target must be included before the mbed-os subdir.
```
?

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,21 +1,77 @@
...
+target_link_libraries(main  mbed-os) 
+
+### link user library (if needed)
+#target_link_libraries(main YourLibrary) 
+
+### End of block
+######################################################################################################
+
+### Set post build (necessary everytime)
+mbed_set_post_build(main)
```
Comment (multiplemonomials) [inline CMakeLists.txt]: Nitpick: can we move this up below the `add_executable()` call?  Really each `add_executable()` should be paired up with an `mbed_set_post_build()` call.  Note that it's perfectly ok to call mbed_set_post_build() before target_link_libraries().

File: README.md [hunk 5]
```diff
@@ -10,9 +10,11 @@ This repo includes three different custom targets you can use based on the `NUCL
 
 ## How to set up this project:
 
-1. Clone it to your machine.  Don't forget to use `--recursive` to clone the submodules: `git clone --recursive https://github.com/mbed-ce/mbed-ce-custom-targets.git`
+1. Clone it to your machine.  Don't forget to use `--recursive` to clone the submodules: `git clone --recursive https://github.com/mbed-ce/mbed-ce-custom-targets.git`and update MbedOS with `git submodule update --remote mbed-os` (call it in project folder)
```
Comment (multiplemonomials) [inline README.md]: Nitpick: "(call it in project folder)" -> "(run this command from the mbed-ce-custom-targets folder)"

File: README.md [hunk 6]
```diff
@@ -10,9 +10,11 @@ This repo includes three different custom targets you can use based on the `NUCL
...
 2. Set up the GNU ARM toolchain (and other programs) on your machine using [the toolchain setup guide](https://github.com/mbed-ce/mbed-os/wiki/Toolchain-Setup-Guide).
-3. Set up the CMake project for editing.  Use one of the target names listed above.  We have three ways to do this:
+3. Set up the CMake project for editing. We have three ways to do this:
     - On the [command line](https://github.com/mbed-ce/mbed-os/wiki/Project-Setup:-Command-Line)
     - Using the [CLion IDE](https://github.com/mbed-ce/mbed-os/wiki/Project-Setup:-CLion)
     - Using the [VS Code IDE](https://github.com/mbed-ce/mbed-os/wiki/Project-Setup:-VS-Code)
+4. Use one of the target names listed above and choose an [upload method](https://github.com/mbed-ce/mbed-os/wiki/Upload-Methods) according to your requirments. Default settings is target L452RE_SIMPLE and upload method STM32CUBE.
```
Comment (multiplemonomials) [inline README.md]: typo: requirments

File: cmake-variants.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  # Fill name of your targer which has to corespond to
+  # target name in custom_targets.json5 file
+  default: L452RE_SIMPLE
+  choices:
+    YOUR_MBED_TARGET:
+      short: L452RE_SIMPLE
+      settings:
+        MBED_TARGET: L452RE_SIMPLE   
+        # Fill your upload method
+        # variants are: NONE, MBED, STLINK, JLINK, PYOCD, OPENOCD, STM32CUBE
```
Comment (multiplemonomials) [inline cmake-variants.yaml]: LinkServer: 🥺

File: custom_targets/upload_method_cfg/L452RE_CUSTOM_CLOCK.cmake [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+# General config parameters
+# -------------------------------------------------------------
+set(UPLOAD_METHOD_DEFAULT OPENOCD)
+
+# Config options for OPENOCD
+# -------------------------------------------------------------
+set(OPENOCD_UPLOAD_ENABLED TRUE)
+set(OPENOCD_CHIP_CONFIG_COMMANDS
+    -f ${CMAKE_CURRENT_LIST_DIR}/openocd_cfgs/stm32l452re.cfg)
```
Comment (multiplemonomials) [inline custom_targets/upload_method_cfg/L452RE_CUSTOM_CLOCK.cmake]: Wait, I don't think this will work, because stm32l452re.cfg isn't actually in this repo

**Review by multiplemonomials — COMMENTED** [review]
Nice work on this, thanks for adding this stuff!  Will def make it easier for people to set up upload methods.

Also sorry if I had to get a little nit-picky on some of the comments, I just really want to make sure that it's as clear to people as possible.

File: CMakeLists.txt (hunk 2 above)
Comment (JohnK1987) [reply CMakeLists.txt]: The plan was create a one file which will contain all basic informations for custom target project but also for regular project (could be also part of blinky example). So this should be just a mark of a sub-block under the including block related only to the custom target project settings.

Does the new description below look better?

> This part is dedicated for custom targets only! Settings below activate targets from custom_targets.json5 
and upload method config, otherwise functions below should be commented.

File: CMakeLists.txt (hunk 1 above)
Comment (JohnK1987) [reply CMakeLists.txt]: I had the link there but was removed before this PR :)

About "of course you can do it by yourself directly via cmake in this file" this is just or info. I saw CMake experienced user use this way automatically.
I suggest to add just the link and the rest into the wiki page. What do you think?

File: CMakeLists.txt (hunk 3 above)
Comment (JohnK1987) [reply CMakeLists.txt]: Of course

File: CMakeLists.txt (hunk 4 above)
Comment (JohnK1987) [reply CMakeLists.txt]: As you wish.

File: cmake-variants.yaml (hunk 7 above)
Comment (JohnK1987) [reply cmake-variants.yaml]: Good point, it was my old helper line and i forgot also Picotool and Arduino bossac.

File: custom_targets/upload_method_cfg/L452RE_CUSTOM_CLOCK.cmake (hunk 8 above)
Comment (JohnK1987) [reply custom_targets/upload_method_cfg/L452RE_CUSTOM_CLOCK.cmake]: Yes, that is good catch!  I will try to fix this somehow.

**Comment by JohnK1987** [comment]
> Also sorry if I had to get a little nit-picky on some of the comments, I just really want to make sure that it's as clear to people as possible.

It is not necessary to apologize, that is why we do it like this. I am also aware my english skill is no so high so every correction are welcomed.

Please be so kind and review the changes i made.

File: CMakeLists.txt [hunk 9]
```diff
@@ -1,21 +1,76 @@
...
+### include app.cmake (necessary everytime) ###
 include(mbed-os/tools/cmake/app.cmake)
 
-# Include folder with the custom targets.
-# This needs to be done before add_subdirectory(mbed-os).
+######################################################################################################
+### Block of including project folders
+
+#[[ If using a custom target, the subdirectory containing the custom target must be included before 
+    the mbed-os subdir, otherwise should be the next line commented]]
```
Comment (multiplemonomials) [inline CMakeLists.txt]: "otherwise the next line should be commented"

File: CMakeLists.txt [hunk 10]
```diff
@@ -1,21 +1,76 @@
...
+
+#[[ Here you can set path for custom upload config .cmake (optional example) ]]
+# set(CUSTOM_UPLOAD_CFG_PATH ${CMAKE_SOURCE_DIR}/${MBED_TARGET}/${MBED_TARGET}.cmake)
+
+#[[ Note: For custom target you need also an upload method and we have few options how you can do that
+    - use the variable CUSTOM_UPLOAD_CFG_PATH above
+    - use default expected path for custom targets upload methods where you create your own config
+      MY_PROJECT/custom_targets/upload_method_cfg
+    - of course you can do it by yourself directly via cmake in this file
+   Fro more visit https://github.com/mbed-ce/mbed-os/wiki/Upload-Methods ]]
```
Comment (multiplemonomials) [inline CMakeLists.txt]: Typo, Fro -> For

**Comment by multiplemonomials** [comment]
Just two more little fixes then good to go!
